### PR TITLE
Neon encoder/decoder

### PR DESCRIPTION
**Pull request by KWillets**
ARM NEON intrinsics for encoding and decoding.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -4,8 +4,12 @@
 .SUFFIXES: .cpp .o .c .h
 
 PROCESSOR:=$(shell uname -m)
-ifeq ($(PROCESSOR), aarch64) 
-# for ARM processors
+
+ifeq ($(PROCESSOR), aarch64)
+# for 64-bit ARM processors
+CFLAGS = -fPIC -std=c99 -O3 -Wall -Wextra -pedantic -Wshadow -D__ARM_NEON__
```
Comment (KWillets) [inline Makefile]: __ARM_NEON__ doesn't seem to be appearing on its own, so it's defined here.

File: src/neon_poc.c [hunk 2]
```diff
@@ -0,0 +1,86 @@
+// usage(pi):  gcc -mfpu=neon -mfloat-abi=hard -std=c99 neon_poc.c
```
Comment (KWillets) [inline src/neon_poc.c]: This will be removed.

File: src/streamvbyte.c [hunk 3]
```diff
@@ -76,6 +77,88 @@ static uint8_t *svb_encode_scalar(const uint32_t *in,
...
+  // lane code is 3 - (saturating sub) (clz(data)/8)
+  uint32x4_t clzbytes = vshrq_n_u32(vclzq_u32(data), 3);
+  uint32x4_t lanecodes = vqsubq_u32(vdupq_n_u32(3), clzbytes);
+
+  // nops
+  uint8x16_t lanebytes = vreinterpretq_u8_u32(lanecodes);
+  uint8x8x2_t twohalves = {{vget_low_u8(lanebytes), vget_high_u8(lanebytes)}};
+
+  // shuffle lsbytes into two copies of an int
+  uint8x8_t lobytes = vtbl2_u8(twohalves, gatherlo);
```
Comment (KWillets) [inline src/streamvbyte.c]: This can be changed to 16-byte on aarch64.

**Comment by lemire** [comment]
Thanks. On my todo... if not today then tomorrow.

File: src/streamvbyte.c [hunk 4]
```diff
@@ -76,6 +77,88 @@ static uint8_t *svb_encode_scalar(const uint32_t *in,
...
+  uint32_t code = codeAndLength[0] >> 24;
+  size_t length = 4 + (codeAndLength[1] >> 24);
+
+  // shuffle in 8-byte chunks
+  uint8x16_t databytes = vreinterpretq_u8_u32(data);
+  //  uint8x8x2_t datahalves = {{vget_low_u8(databytes), vget_high_u8(databytes)}};
+
+  uint8x16_t encodingShuffle = vld1q_u8((uint8_t *) &encodingShuffleTable[code]);
+
+  //  vst1_u8(outData, vtbl2_u8(datahalves, vget_low_u8(encodingShuffle)));
```
Comment (KWillets) [inline src/streamvbyte.c]: todo:  possibly ifdef the 8-byte shuffles for < aarch64.

File: src/streamvbyte.c [hunk 5]
```diff
@@ -155,11 +220,22 @@ size_t streamvbyte_encode(uint32_t *in, uint32_t count, uint8_t *out) {
   uint8_t *keyPtr = out;
   uint32_t keyLen = (count + 3) / 4;  // 2-bits rounded to full byte
   uint8_t *dataPtr = keyPtr + keyLen; // variable byte data after all keys
-#ifdef __AVX__
-  return svb_encode_vector(in, keyPtr, dataPtr, count) - out;
-#else
-  return svb_encode_scalar(in, keyPtr, dataPtr, count) - out;
+
+#if defined(__AVX__) || defined(__ARM_NEON__)
```
Comment (KWillets) [inline src/streamvbyte.c]: This is an attempt to unify the interface for vector-based encoding in 4-word quads.  Another issue/branch should address breaking these out in an inline/static-friendly way.

File: src/streamvbyte.c [hunk 6]
```diff
@@ -311,12 +387,23 @@ const uint8_t *svb_decode_avx_simple(uint32_t *out,
...
+
   const uint8_t *keyPtr = in;               // full list of keys is next
   uint32_t keyLen = ((count + 3) / 4);      // 2-bits per key (rounded up)
   const uint8_t *dataPtr = keyPtr + keyLen; // data starts at end of keys
+
 #ifdef __AVX__
-  return svb_decode_avx_simple(out, keyPtr, dataPtr, count) - in;
-#else
-  return svb_decode_scalar(out, keyPtr, dataPtr, count) - in;
+  dataPtr = svb_decode_avx_simple(out, keyPtr, dataPtr, count);
```
Comment (KWillets) [inline src/streamvbyte.c]: This is also an attempt to inline the vector decoding and scalar tail-processing.  The AVX and NEON versions have different vector lengths, so we figure the lengths in each version.

**Comment by KWillets** [comment]
Thanks, it's interesting to see how SIMD works in practice, and this was just big enough to learn on.

**Pull request merged**